query/expression.go

```diff
@@ -34,7 +34,31 @@ type EvaluationContext struct {
 	API          api.API          // Api to obtain metadata from
 	Timerange    api.Timerange    // Timerange to fetch data from
 	SampleMethod api.SampleMethod // SampleMethod to use when up/downsampling to match the requested resolution
-	Predicate    api.Predicate
+	Predicate    api.Predicate    // Predicate to apply to TagSets prior to fetching
+	FetchLimit   fetchCounter     // A limit on the number of fetches which may be performed
+}
+
+// fetchCounter is used to count the number of fetches remaining in a thread-safe manner.
+type fetchCounter struct {
+	count  *int
+	ticket chan struct{}
+}
+
+func NewFetchCounter(n int) fetchCounter {
+	f := fetchCounter{
+		count:  &n,
+		ticket: make(chan struct{}),
+	}
+	f.ticket <- struct{}{}
+	return f
+}
+
+func (c fetchCounter) Consume(n int) bool {
+	<-c.ticket
+	*c.count -= n
+	r := *c.count < 0
+	c.ticket <- struct{}{}
+	return r
 }
```

Comment:
This function performs the decrement / compare in a thread-safe manner (it can't be interrupted because the channel will block for all other threads). So now it's safe to make the rest of evaluation concurrent without causing any problems here.